### PR TITLE
Revamp SPMD guide

### PR DESCRIPTION
**Pull request by tengyifei**
I was going to document some device ID assignment tricks and then I realized in order to explain that we need to first explain what is a device mesh. The existing documentation is very poor so I'm adding an in-depth guide in this PR.

While going over the APIs used I noticed the API documentation and typing also has some problems so I fixed those while I'm here.

### CODE REVIEW

**Review by qihqi — APPROVED** [review]

**Review by yaoshiang — APPROVED** [review]
Approved with some optional nits. Great work on cleaning this up!

File: docs/source/perf/spmd_basic.md [hunk 1]
```diff
@@ -61,35 +65,93 @@ SPMD with other distributed libraries.
...
 
 ``` python
     >>> mesh.shape()
     OrderedDict([('data', 4), ('model', 1)])
+    
+    >>> mesh.get_logical_mesh()
+    array([[0], [1], [2], [3]])
+    
+    # Details about these 4 TPUs
+    >>> xr.global_runtime_device_attributes()
```
Comment (pgmoka) [inline docs/source/perf/spmd_basic.md:111]: It might be useful to define what `xr` is. Maybe in an earlier comment add `import torch_xla.runtime as xr`

Comment (pgmoka) [reply docs/source/perf/spmd_basic.md:111]: It is on a previous step. I didn't spot it previously

File: docs/source/perf/spmd_basic.md [hunk 2]
```diff
@@ -114,3 +193,7 @@ xs.mark_sharding( t2, mesh, (('data', 'model'),))
     topics](https://github.com/pytorch/xla/blob/master/docs/spmd_advanced.rst)
 4.  [Spmd Distributed
     Checkpoint](https://github.com/pytorch/xla/blob/master/docs/spmd_distributed_checkpoint.rst)
```
Comment (pgmoka) [inline docs/source/perf/spmd_basic.md:195]: These seem to be outdated. we should use:
- https://github.com/pytorch/xla/blob/6804c65f3bb464a899b84e39830fe09cbb4d350e/docs/source/perf/spmd_advanced.md#L4 
-  https://github.com/pytorch/xla/blob/6804c65f3bb464a899b84e39830fe09cbb4d350e/docs/source/perf/spmd_distributed_checkpoint.md

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 3]
```diff
@@ -21,13 +21,20 @@
 
 from torch.amp import custom_fwd, custom_bwd
 
+PartitionSpec = Tuple[Union[Tuple[Union[int, str], ...], int, str, None], ...]
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:24]: Ah yes!

And I have to remember that the `A | B` syntax is only supported in Python 3.10 so we can't use that 🤣

File: docs/source/perf/spmd_basic.md [hunk 4]
```diff
@@ -61,35 +65,93 @@ SPMD with other distributed libraries.
 
 ### Mesh
 
-For a given cluster of devices, a physical mesh is a representation of
-the interconnect topology.
+The SPMD programming model revolves around the concept of a "device mesh", also
+called "mesh" for short. A device mesh is a logical N-dimensional arrangement of
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:69]: Got it. Reworded.

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 5]
```diff
@@ -21,13 +21,20 @@
...
+"""
+A partition spec is a tuple where each element describes how the corresponding
+input tensor dimension is sharded across the device mesh.
+"""
+
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:29]: > First, usually for docstrings, the text starts immediately after the first triple quote, [...]

fixed

yeah lol it's not trivial to build the docs locally. i will for now stay satisfied when vscode can give me nice hover-over documentation for PartitionSpec

File: docs/source/perf/spmd_basic.md (hunk 2 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:195]: Fixed

File: docs/source/perf/spmd_basic.md [hunk 6]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
 
 ### Mesh
 
-For a given cluster of devices, a physical mesh is a representation of
-the interconnect topology.
+The SPMD programming model is built around the concept of a device mesh, commonly
+referred to as a mesh. A device mesh is a logical N-dimensional arrangement of
+compute devices (e.g. TPU cores) where the axes are used as shorthands to
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:70]: What does "the axes are used as shorthands to define sharded computation" mean?

File: docs/source/perf/spmd_basic.md [hunk 7]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+define sharded computation. It is logical in the sense that the mesh shape does
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:71]: It is logical ==> The device mesh shape does not necessarily reflect the physical network layout.

File: docs/source/perf/spmd_basic.md [hunk 8]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+not necessarily have to reflect the physical network layout. You could create
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:72]: You can create differently shaped meshes over the same set of physical devices.

File: docs/source/perf/spmd_basic.md [hunk 9]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+different kinds of device meshes over the same set of physical devices. For
+example, a 512-core TPU slice could be treated as a 3D mesh of 16×16×2, a 2D mesh
+of 32×16, or a 1D mesh of 512, depending on how we want to partition tensors. You
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:75]: ... depending on how you want to partition tensors.

File: docs/source/perf/spmd_basic.md [hunk 10]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+may use the `Mesh` class to create a device mesh.
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:76]: You use the `Mesh` class to create a device mesh.

File: docs/source/perf/spmd_basic.md [hunk 11]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+  of physical devices in the environment.
+- `device_ids` specifies the ordering of physical devices within the mesh in row
+  major order. It is always a flat numpy array of integers ranging from `0` to
+  `num_devices - 1`, where `num_devices` is the total number of devices in the
+  environment. Each ID indexes into the list of devices in
+  `xr.global_runtime_device_attributes()`.
+  The simplest ordering is `np.array(range(num_devices))`, but tuning the
+  device ordering and the mesh shape to utilize the underlying physical
+  interconnect can improve efficiency.
+- Users are also encouraged to give each mesh axis a name. Later we can
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:96]: Best practice is to give each mesh axis a name. You can then shard each dimension of tensors over specific mesh axes ....

File: docs/source/perf/spmd_basic.md [hunk 12]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+  shard each dimension of tensors over specific mesh axes to achieve the
+  desired parallelization. In the above example, the first mesh dimension is the
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:98]: In the previous example, ...

File: docs/source/perf/spmd_basic.md [hunk 13]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+mesh = Mesh(device_ids, mesh_shape, ('data', 'model'))
+```
 
-1.  `mesh_shape` is a tuple that will be multiplied to the total number
-    of physical devices.
-2.  `device_ids` is almost always `np.array(range(num_devices))`.
-3.  Users are also encouraged to give each mesh dimension a name. In the
-    above example, the first mesh dimension is the `data` dimension and
-    the second mesh dimension is the `model` dimension.
+- `mesh_shape` is a tuple whose elements must multiply to the total number
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:86]: Would it be correct to say "mesh_shape is a tuple whose elements describe the size of each axis (dimension?) of the mesh"? Then we can say "multiplying the elements of the tuple gives you the total number of physical devices"?

File: docs/source/perf/spmd_basic.md [hunk 14]
```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
...
+partition_spec = ('data', 'model')
+xs.mark_sharding(t, mesh, partition_spec)
+```
+
+`partition_spec` has the same rank as the input tensor. Each dimension
 describes how the corresponding input tensor dimension is sharded across
 the device mesh. In the above example tensor `t`'s fist dimension is
-being sharded at `data` dimension and the second dimension is being
-sharded at `model` dimension.
+being sharded over the `data` dimension and the second dimension is being
```
Comment (mikegre-google) [inline docs/source/perf/spmd_basic.md:149]: In the above example, the first dimension of tensor `t` is sharded over the `data` dimension and the second dimension is sharded over the `model` dimension.

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 15]
```diff
@@ -16,24 +16,33 @@
...
+from typing import Union, Sequence, Any, Optional
 from enum import IntEnum
 
 from torch.amp import custom_fwd, custom_bwd
 
+PartitionSpec = tuple[Union[tuple[Union[int, str], ...], int, str, None], ...]
+"""PartitionSpec describes the sharding of a tensor.
+
+Specifically, it is a tuple where each element references one or more device mesh
+axes, and instructs to shard the corresponding dimension of the input tensor over
```
Comment (mikegre-google) [inline torch_xla/distributed/spmd/xla_sharding.py:28]: Just to make sure I understand what this is saying:  Is the first dimension of the input tensor sharded across the first element of the tuple, and so on? If so, we can say that. "... it is a tuple of one or more device mesh axes that describes how to shard the input tensor. For example, the first element of the tuple is sharded across the axis described in the first element of the tuple and so on."

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 16]
```diff
@@ -16,24 +16,33 @@
...
+"""
+
 
 class Mesh:
   """Describe the logical XLA device topology mesh and the underlying resources.
 
   Args:
-    device_ids (Union[np.ndarray, List]): A raveled list of devices (IDs) in a custom order. The list is reshaped
-        to an `mesh_shape` array, filling the elements using C-like index order.
+    device_ids: A raveled list of devices (IDs) in a custom order.
```
Comment (mikegre-google) [inline torch_xla/distributed/spmd/xla_sharding.py:37]: Instead of "raveled" can we say "flattened"? I'm unfamiliar with the term, but I see there is a numpy function called `ravel`. It's best not to use a function as a verb, but rather use a verb that describes what the function is doing.

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 17]
```diff
@@ -16,24 +16,33 @@
...
+        The list is reshaped to an `mesh_shape` array, filling the elements using C-like index order.
+        Each ID indexes into the list of devices returned by `xr.global_runtime_device_attributes()`.
 
-    mesh_shape (Tuple[int, ...]): A int tuple describing the logical topology shape
+    mesh_shape: A int tuple describing the logical topology shape
         of the device mesh, and each element describes the number of devices in
```
Comment (mikegre-google) [inline torch_xla/distributed/spmd/xla_sharding.py:42]: ... of the device mesh. Each element describes ....

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 18]
```diff
@@ -16,24 +16,33 @@
...
         the corresponding axis.
 
-    axis_names (Tuple[str, ...]): A sequence of resource axis names to be assigned to the dimensions
+    axis_names: A sequence of resource axis names to be assigned to the dimensions
```
Comment (mikegre-google) [inline torch_xla/distributed/spmd/xla_sharding.py:45]: A sequence of axis names, each name is assigned to the corresponding dimension of the `devices` argument.

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 19]
```diff
@@ -495,10 +503,10 @@ def enable_manual_sharding(t: Union[torch.Tensor, XLAShardedTensor],
...
 def disable_manual_sharding(t: Union[torch.Tensor, XLAShardedTensor],
-                            partition_spec: Tuple[Union[Tuple, int, str, None]],
+                            partition_spec: PartitionSpec,
                             full_shape: torch.Size,
                             *,
-                            mesh: Mesh = None) -> XLAShardedTensor:
+                            mesh: Optional[Mesh] = None) -> XLAShardedTensor:
   """
   This API disables manual sharding for the given tensor. The partition_spec and full_shape are used to construct the
   output tensor as if the input tensor has not been manual sharded.
```
Comment (mikegre-google) [inline torch_xla/distributed/spmd/xla_sharding.py:512]: ... as if the input tensor has not been manually sharded.

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 20]
```diff
@@ -524,7 +530,7 @@ def mark_sharding(
 
         mesh (Mesh): describes the logical XLA device topology and the underlying device IDs.
 
-        partition_spec (Tuple[Tuple, int, str, None]): A tuple of device_mesh dimension index or
+        partition_spec (PartitionSpec): A tuple of device_mesh dimension index or
             `None`. Each index is an int, str if the mesh axis is named, or tuple of int or str.
```
Comment (mikegre-google) [inline torch_xla/distributed/spmd/xla_sharding.py:534]: This makes it sound like there are two options: 1 each index is an int, str tuple or 2 each index is an int, str tuple. But both options are the same. Am I misunderstanding?

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 21]
```diff
@@ -524,7 +530,7 @@ def mark_sharding(
...
             This specifies how each input rank is sharded (index to mesh_shape) or replicated (None).
             When a tuple is specified, the corresponding input tensor axis will be sharded along all
```
Comment (mikegre-google) [inline torch_xla/distributed/spmd/xla_sharding.py:536]: This sentence is not complete. The corresponding input tensor axis will be sharded along all ... ???

**Review by mikegre-google — COMMENTED** [review]
Much better descriptions. I have few questions and suggestions..

File: docs/source/perf/spmd_basic.md (hunk 6 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:70]: Expanded to "the axes are used as shorthands to define the devices that participate in a sharded computation".

Basically each axis of the mesh will let you select a bunch of devices. And later when you say I want to shard a tensor over this axis, that means shard (distribute pieces of) the tensor over those devices. So the axis is a shorthand.

Wondering if you had clearer ideas?

File: docs/source/perf/spmd_basic.md (hunk 7 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:71]: done

File: docs/source/perf/spmd_basic.md (hunk 8 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:72]: done

File: docs/source/perf/spmd_basic.md (hunk 10 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:76]: done

File: docs/source/perf/spmd_basic.md (hunk 11 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:96]: done

File: docs/source/perf/spmd_basic.md (hunk 12 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:98]: done

File: docs/source/perf/spmd_basic.md (hunk 13 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:86]: yes

Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:86]: i think i'll prefer to use the word "axis" whenever it comes to the device mesh, and "dimension" whenever it comes to tensors. just makes it easier to tell things apart contextually.

File: docs/source/perf/spmd_basic.md (hunk 14 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md:149]: done

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 15 above)
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:28]: yes

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 16 above)
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:37]: yes. i didn't write this originally. and i was also confused by this word

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 17 above)
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:42]: done

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 18 above)
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:45]: done

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 19 above)
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:512]: done

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 20 above)
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:534]: Each element can be:

- int (refer to mesh axis by index)
- str (refer to mesh axis by name)
- a tuple of the above

In the third case, the corresponding dimension of the tensor will be sharded over multiple mesh axes.

..now that i wrote this out i might as well rewrite this whole confusing section too..

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 21 above)
Comment (tengyifei) [reply torch_xla/distributed/spmd/xla_sharding.py:536]: hmmm, it's pretty complete if i view the whole file

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 20 above)
Comment (mikegre-google) [reply torch_xla/distributed/spmd/xla_sharding.py]: Ah, ok! Great job explaining this!

**Comment by mikegre-google** [comment]
Looks good.  Only one further suggestion...

File: docs/source/perf/spmd_basic.md (hunk 6 above)
Comment (tengyifei) [reply docs/source/perf/spmd_basic.md]: Basically, rather than saying "do the all-gather for device 2,4,6,8,10,12.......", we can say "do the all-gather over the fsdp axis"

**Review by mikegre-google — APPROVED** [review]
Looks good!

**Comment by bhavya01** [comment]
Thanks for making the change. The reference to JAX guide will be very helpful!
